notebooks/Batchelor.ipynb

```diff
@@ -0,0 +1,348 @@
+    "    u_bc2 = Function(V_v)\n",
+    "    u_bc2.interpolate(u_0)\n",
+    "    bc2 = dirichletbc(u_bc2, dofs_D_2, V.sub(0))\n",
+    "\n",
+    "    # define boundary value for Dirichlet boundary condition (61)\n",
+    "    def boundary_D_3(x):\n",
+    "        return np.logical_or(np.isclose(x[0], 1), np.isclose(x[1], 1))\n",
+    "\n",
+    "    dofs_D_3 = locate_dofs_geometrical((V.sub(0), V_v), boundary_D_3)\n",
+    "    u_bc3 = Function(V_v)\n",
+    "    u_bc3.interpolate(u_exact_np)\n",
+    "    bc3 = dirichletbc(u_bc3, dofs_D_3, V.sub(0))\n",
+    "\n",
+    "    def boundary_D_4(x):\n",
+    "        return np.logical_and(np.isclose(x[0], 0), np.isclose(x[1], 0))\n",
+    "\n",
+    "    dofs_D_4 = locate_dofs_geometrical((V.sub(1), V_p), boundary_D_4)\n",
+    "    u_bc4 = Function(V_p)\n",
+    "    u_bc4.x.array[:] = 0\n",
+    "    bc4 = dirichletbc(u_bc4, dofs_D_4, V.sub(1))\n",
+    "\n",
+    "    # assemble weak form\n",
+    "    K = ufl.inner(ufl.sym(ufl.grad(v_t)), ufl.sym(ufl.grad(trial_v))) * ufl.dx\n",
+    "    G = -ufl.div(v_t)*trial_p*ufl.dx\n",
+    "    D = -p_t*ufl.div(trial_v)*ufl.dx\n",
+    "    S = K + G + D\n",
+    "\n",
+    "    zeroes = Constant(domain, default_scalar_type(0.0))\n",
+    "    L = zeroes*p_t*ufl.dx\n",
+    "\n",
+    "    # assemble problem solver and solve\n",
+    "    problem = LinearProblem(S, L, bcs=[bc1, bc2, bc3, bc4], petsc_options={\"ksp_type\": \"preonly\", \"pc_type\": \"lu\", \"pc_factor_mat_solver_type\": \"mumps\"})\n",
+    "    uh = problem.solve()\n",
+    "\n",
+    "    vh = uh.sub(0).collapse()\n",
+    "\n",
+    "    return vh"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "de15d3f6-6e1d-41d3-8f56-bcf76a7b1a59",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def error(vh):\n",
+    "    x = ufl.SpatialCoordinate(vh.function_space.mesh)\n",
+    "    vex = u_exact(x)\n",
+    "    error_L2 = assemble_scalar(form((vh - vex)**2 * ufl.dx))\n",
+    "    error_L2 = np.sqrt(MPI.COMM_WORLD.allreduce(error_L2, op=MPI.SUM))\n",
+    "\n",
+    "    xs = np.linspace(0, 1, 1001)  # third argument is arbitrary\n",
+    "    xgrid, ygrid = np.meshgrid(xs,xs)\n",
+    "    xyzs = np.stack((xgrid.flatten(), ygrid.flatten(), np.zeros_like(xgrid.flatten())), axis=1)\n",
+    "    cells = get_first_cells(xyzs, vh.function_space.mesh)\n",
+    "    vh_vals = vh.eval(xyzs, cells)[:,0]\n",
+    "\n",
+    "    vex_vals = u_exact_np(xyzs.transpose())\n",
+    "    error_max = np.max(np.abs(vh_vals - vex_vals))\n",
+    "    error_max = MPI.COMM_WORLD.allreduce(error_max, op=MPI.MAX)\n",
+    "    \n",
+    "    return error_L2, error_max"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a035fd02-45c8-4a9b-8bf1-52f0e300aafc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# define a utility function to help with visualization\n",
```

Comment:
If error_max isn't calculated anymore I don't think `get_first_cells` is necessary so this entire cell could go.